File: bindings/python/examples/validator_operations.py
# Copyright (c) 2025 IOTA Stiftung
# SPDX-License-Identifier: Apache-2.0

"""
Example: Validator Operations

Demonstrates querying and analyzing validators.
"""

from lib.iota_sdk import *
import asyncio


async def main():
    client = GraphQlClient.new_devnet()

    print("=== Validator Operations Example ===\n")

    # Step 1: Query validators
    print("1. Querying active validators...")
    validators = await client.active_validators(None, {})
    print(f"   Found {len(validators['data'])} active validators\n")

    # Step 2: Display validator info
    print("2. Validator Information:")
    for idx, validator in enumerate(validators['data'][:5], 1):
        print(f"\n   Validator #{idx}:")
        if 'name' in validator and validator['name']:
            print(f"     Name: {validator['name']}")
        print(f"     Address: {validator['address']['address']}")
        print(f"     Voting Power: {validator['voting_power']}")

        if 'nextEpochStakeApy' in validator and validator['nextEpochStakeApy']:
            print(f"     Staking APY: {validator['nextEpochStakeApy']:.2f}%")

        print(f"     Pending Stake: {validator['pendingStake']} MIST")
        print(f"     Pending Withdraw: {validator['pendingWithdraw']} MIST")

    # Step 3: Best practices
    print("\n3. Validator Selection Criteria:")
    print("   - Check commission rates")
    print("   - Review voting power")
    print("   - Consider uptime")
    print("   - Evaluate staking APY")
    print("   - Assess reputation\n")

    # Step 4: Staking workflow
    print("4. Staking Workflow:")
    print("   a. Choose validator")
    print("   b. Prepare transaction")
    print("   c. Sign and submit")
    print("   d. Monitor rewards")
    print("   e. Unstake when needed\n")

    print("Validator operations example completed!")


if __name__ == "__main__":
    asyncio.run(main())

File: bindings/kotlin/examples/ValidatorOperations.kt
// Copyright (c) 2025 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package iota.sdk.examples

import iota.sdk.GraphQlClient

suspend fun main() {
    val client = GraphQlClient.newDevnet()

    println("=== Validator Operations Example ===\n")

    // Step 1: Query validators
    println("1. Querying active validators...")
    val validators = client.activeValidators(null, emptyMap())
    println("   Found ${validators.data.size} active validators\n")

    // Step 2: Display validator information
    println("2. Validator Information:")
    validators.data.take(5).forEachIndexed { idx, validator ->
        println("\n   Validator #${idx + 1}:")
        validator.name?.let { println("     Name: $it") }
        println("     Address: ${validator.address.address}")
        println("     Voting Power: ${validator.votingPower}")

        validator.nextEpochStakeApy?.let { apy ->
            println("     Staking APY: ${"%.2f".format(apy)}%")
        }

        println("     Pending Stake: ${validator.pendingStake} MIST")
        println("     Pending Withdraw: ${validator.pendingWithdraw} MIST")
    }

    // Step 3: Best practices
    println("\n3. Validator Selection Criteria:")
    println("   - Check commission rates")
    println("   - Review voting power distribution")
    println("   - Consider uptime and performance")
    println("   - Evaluate staking APY")
    println("   - Assess community reputation\n")

    // Step 4: Staking workflow
    println("4. Staking Workflow:")
    println("   a. Choose a validator")
    println("   b. Prepare staking transaction")
    println("   c. Sign and submit transaction")
    println("   d. Monitor staking rewards")
    println("   e. Unstake when needed\n")

    println("Validator operations example completed!")
    println("See also: Stake.kt, Unstake.kt")
}
